python-backend/app/auth.py

```diff
@@ -0,0 +1,65 @@
+import re
+
+from cached_property import cached_property
+from flask import g, request
+from uuid import UUID
+from typing import Mapping, Optional, Set
+from .api.utils.include.user_info import User
+from .api.users.minespace.models.minespace_user import MinespaceUser
+from app.api.utils.access_decorators import MINESPACE_PROPONENT
+
+
+class UserSecurity(object):
+    def __init__(self, user_id: int, access: Optional[Set[UUID]] = None):
+        self.access = access or {}
+        self.user_id = user_id
+
+    def __repr__(self):
+        return "<{} mine_ids={}>".format(type(self).__name__, self.mine_ids)
+
+    def is_restricted(self):
+        return get_user_is_proponent()
+
+    @cached_property
+    def mine_ids(self) -> Set[UUID]:
+        if not self.user_id:
+            return None
+
+        return get_mine_access()
+
+    def get_permission(self, mine_id: UUID):
+        return self.access.get(mine_id)
+
+
+def get_mine_access():
+    user = get_current_user()
+    return list(x.mine_guid for x in user.mines)
+
+
+def get_current_user():
+    rv = getattr(g, 'current_user', None)
+    if rv == None:
+        email = get_user_email()
+        rv = MinespaceUser.query.unbound_unsafe().filter_by(email=email).filter_by(
+            deleted_ind=False).first()
+        g.current_user = rv
+    return rv
+
+
+def get_user_is_proponent():
+    token_data = User().get_user_raw_info()
+    is_proponent = MINESPACE_PROPONENT in token_data["realm_access"]["roles"]
```

Comment:
If we're accessing the keys on `token_data` like this, we probably want a `try` and catch a key error.

Otherwise, a `.get()` call; possibly with a `None` check to ensure desired behaviour in returning `is_proponent`

Reply:
Yeah that's a good idea. I'm in the process of adding some handling around whether the token exists or not as it throws an error if there is no token. I'll add some protection here as well.